### PR TITLE
Refactor AWS Role Creation

### PR DESCRIPTION
**Pull request by cat5inthecradle**
Create two new roles to replace the current GoogleUser role.

* Engineer_ReadOnly is intended to be the preferred role for 90% of the reason we log into aws - just to look at things. Choose this role for peace of mind.
* Engineer_FullAccess is nearly identical to the current GoogleUser role, minus the PassRole permission, which I need to research more.

I'd like to merge this and enable them for Infra engineers to begin testing them.

## Deployment strategy

manually deployed via rake command by an AWS Admin.

This will break access for Contributors and RED Team until we update their google accounts and they update their local aws config. Our AWS Access document would need to be updated with new ARN's.

## PR Checklist:

- [ ] Tests provide adequate coverage
- [ ] Privacy and Security impacts have been assessed
- [ ] Code is well-commented
- [ ] New features are translatable or updates will not break translations
- [ ] Relevant documentation has been added or updated
- [ ] User impact is well-understood and desirable
- [ ] Pull Request is labeled appropriately
- [ ] Follow-up work items (including potential tech debt) are tracked and linked

### CODE REVIEW

**Review by Hamms — COMMENTED** [review]
How difficult would it be to leverage ERB to reduce the amount of repeated configuration here?

Something like,

```erb
<%
roles = {
  EngineeringReadOnlyRole: {
    RoleName: "Engineering_ReadOnly",
    ManagedPolicyArns: [
      'arn:aws:iam::aws:policy/ReadOnlyAccess'
    ]
  },
  EngineeringFullAccessRole: {
    RoleName: "Engineering_FullAccess",
    ManagedPolicyArns: [
      'arn:aws:iam::aws:policy/PowerUserAccess',
      'arn:aws:iam::aws:policy/IAMReadOnlyAccess'
    ]
  },
}
roles.each do |name, config| %>
  <%= name %>:
    Type: AWS::IAM::Role
    Properties:
      RoleName: <%= config["RoleName"] %>
      AssumeRolePolicyDocument:
        Version: 2012-10-17
        Statement:
          - Effect: Allow
            Principal: {Federated: !Sub "arn:aws:iam::${AWS::AccountId}:saml-provider/Google"}
            Action: sts:AssumeRoleWithSAML
            Condition:
              StringEquals:
                SAML:aud: https://signin.aws.amazon.com/saml
          - Effect: Allow
            Principal: {Federated: accounts.google.com}
            Action: sts:AssumeRoleWithWebIdentity
            Condition:
              StringEquals:
                accounts.google.com:aud: '{{resolve:secretsmanager:admin/google_client_id}}'
              ForAnyValue:StringEquals:
                accounts.google.com:sub: !Ref DeveloperGoogleIDs
      ManagedPolicyArns: <%= role['ManagedPolicyArns'].to_yaml %>
      Policies:
        - PolicyName: "DenyAdminLogs"
          PolicyDocument:
            Version: 2012-10-17
            Statement:
              - Effect: Deny
                Action: logs:*
                Resource: !Sub 'arn:aws:logs:${AWS::Region}:${AWS::AccountId}:log-group:/admin/*'
      Tags:
        - Key: sqlworkbench-team
          Value: codedotorg
<% end %>
```

?

**Comment by sureshc** [comment]
My memory is that engineers need `iam:PassRole` to provision various AWS services (typically via CloudFormation) that need to assume various Roles (that may have been defined in our `IAM` Stack).

**Comment by cat5inthecradle** [comment]
> leverage ERB

@Hamms  I didn't like this at first, but then I decided to go all in and see what happened, and I've come around.

Commit [ab94b4f](https://github.com/code-dot-org/code-dot-org/pull/52407/commits/ab94b4f2f57b80dd4b590e9a414e7af1a148de74) brings all of the human-assuming roles into this pattern.

File: aws/cloudformation/iam.yml.erb [hunk 1]
```diff
@@ -225,15 +273,29 @@ Resources:
             Condition:
               StringEquals:
                 SAML:aud: https://signin.aws.amazon.com/saml
-      ManagedPolicyArns:
-        - !Ref DataAnalystPolicy
-        - !Ref DataAnalystDmsReadOnlyPolicy
-        - !Ref DataAnalystRedshiftPolicy
-        - arn:aws:iam::aws:policy/AmazonRedshiftQueryEditor
-        - arn:aws:iam::aws:policy/AmazonRedshiftQueryEditorV2FullAccess
+          <%if config.GoogleIdsSSMParameter -%>
```
Comment (Hamms) [inline aws/cloudformation/iam.yml.erb]: I believe we'll actually need to use a hash reference with a symbol argument:

```suggestion
          <%if config[:GoogleIdsSSMParameter] -%>
```

For this and all other `config` references, including the ones that I implemented with strings in my original suggestion.

File: aws/cloudformation/iam.yml.erb [hunk 2]
```diff
@@ -211,11 +213,57 @@ Resources:
...
           Value: codedotorg
-  DataAnalystRole:
+
+# Console & CLI Roles
+
+<%
+roles = {
+  EngineeringReadOnlyRole: {
+    RoleName: "Engineering_ReadOnly",
+    GoogleIdsSSMParameter: 'DeveloperGoogleIDs'
```
Comment (Hamms) [inline aws/cloudformation/iam.yml.erb]: nit: a few missing trailing commas
```suggestion
    GoogleIdsSSMParameter: 'DeveloperGoogleIDs',
```

**Review by Hamms — COMMENTED** [review]
Awesome! I wasn't entirely sure of the idea myself, but I'm happy with how many red lines this diff has now.

I'm ignorant of a lot of our specific AWS concerns and requirements here, and I have two ruby syntax nits, but otherwise from a templating perspective this all looks good to me!

**Comment by cat5inthecradle** [comment]
Personally I'm not a huge fan of loops that generate resources - I like to see each resource individually. There's a lot of boilerplate here though, and I want to make it easier to add more roles, so I think this is a good direction.

The `Developer` role above this won't fit this approach because it has policies defined within - that can be fixed by refactoring those into `AWS::IAM::Policy` resources and referencing them in the config like the others. However... I am not convinced that this role is actually _assumed_ by any humans for Console or CLI access. We all use "GoogleUser". I'll leave it as is for now and tackle it as a separate issue.

**Comment by cat5inthecradle** [comment]
I've restored the old roles so current users are not interrupted. We can switch them (via Google and ~/.aws/config) when ready.

**Comment by cat5inthecradle** [comment]
Validated with `VERBOSE=1 bundle exec rake stack:iam:validate RAILS_ENV=production`

Compared generated template with currently deployed template, all changes are as expected.

**Comment by cat5inthecradle** [comment]
deployed.

**Pull request merged**